### PR TITLE
[RLlib] Learner API enhancements and cleanups (prep. for DreamerV3).

### PR DESCRIPTION
**Pull request by sven1977**
Learner API enhancements and cleanups (prep. for DreamerV3). This is a replacement PR for https://github.com/ray-project/ray/pull/35574, which has been closed.

Please read this summary carefully before reviewing, it explains all important changes:

**Learner API**
* optimizer_config has been removed and all settings related to optimizers are now included in the base LearnerHyperparameter class. Purely through AlgorithmConfig, the user can only influence: lr or lr-schedule and basic grad clipping behavior for a single Adam optimizer per module. In all other cases (non-Adam, >1 optimizers per module, c'tor args other than lr, non-piecewise lr schedules), the user **must** override the `configure_optimizers_for_module` method and possibly bring it their own configs and learner HPs.
* A new `register_optimizer()` method has been added to `Learner`, allowing RLlib and users that write custom Learner subclasses to create and register their own optimizers (always under a ModuleID (ALL_MODULES by default) AND an optimizer name (DEFAULT_OPTIMIZER by default)).
* With this `register_optimizer` method,  users now can override either(!) `configure_optimizers` (MARL case where no individual RLModules have their dedicated optimizers) OR `configure_optimizers_for_module` (normal, independent MARL case) and - in there - call the `register_optimizer()` method. This is hence analogous of the just introduced `register_metrics` API.
** This way, we massively simplified the optimizer management and customization options, getting rid of the `_configure_optimizers_for_module_helper` method and typedefs, such as ParamOptimPair(s), NamedParamOptimPair, etc..
** `get_optimizer(module_id=default, optimizer_name=default)`, `get_optimizers_for_module(module_id)` have been added for simpler optimizer access w/o going through the private properties (e.g. `self._named_optimizers`) anymore.
* Renamed `Learner.get_weights` into `Learner.get_module_state()` for clarity and unity.
* Renamed `Learner.get_optimizer_weights` into `Learner.get_optimizer_state` for clarity and unity.

**MARLModules**
* Users can now specify AlgorithmConfig overrides per RLModule within a MultiAgentRLModule (analogous to the old stack's `PolicySpec(config={"lr": 0.001})` syntax). These overrides are compiled only inside the LearnerHyperparameter instances (away from the user) and a new `LearnerHyperparameter.get_hps_for_module` API has been created to extract module-specific settings, e.g. for usage in `compute_loss_for_module()`.
* The module_id_specific HPs are now passed into all of Learner's `..._for_module(module_id, hps, ...)` methods for convenience.
* For complex MARL cases, the user may override `configure_optimizers` directly (and call `register_optimizer` from therein) and define optimizers that operate on the entire MARLModule or on several RLModules therein.

**Test cases**
* The test_torch_learner.py file has been merged with test_learner.py. The resulting test case now tests both frameworks (tf2 and torch) identically.

**Scheduler API**
* Scheduler objects are no longer per-module as this was a design flaw. Learner itself should handle different schedulers for different module_ids.
* A new defaultdict (LambdaDefaultDict) has been added that can assign default values depending on the accessed key (a regular python defaultdict always returns the same default value, regardless of the key). This allows us to simplify the individual algorithms' Learners (e.g. APPOLearner) and NOT have to override `add_module` anymore, just to setup a correct entropy or other Scheduler.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/torch_utils.py [hunk 1]
```diff
@@ -134,9 +134,12 @@ def clip_gradients(
     # Clip by L2-norm (per gradient tensor).
     elif grad_clip_by == "norm":
         for k, v in gradients_dict.copy().items():
-            gradients_dict[k] = (
-                None if v is None else nn.utils.clip_grad_norm_(v, grad_clip)
-            )
+            if v is not None:
```
Comment (sven1977) [inline rllib/utils/torch_utils.py:137]: bug fix

File: rllib/core/learner/torch/torch_learner.py [hunk 2]
```diff
@@ -112,37 +115,6 @@ def compute_gradients(
 
         return grads
 
-    @OverrideToImplementCustomLogic_CallToSuperRecommended
```
Comment (sven1977) [inline rllib/core/learner/torch/torch_learner.py:115]: moved logic to base class. Possible b/c of new enhanced APIs.

File: rllib/core/learner/tf/tf_learner.py [hunk 3]
```diff
@@ -109,28 +112,12 @@ def compute_gradients(
         return grads
 
     @override(Learner)
-    def postprocess_gradients(
```
Comment (sven1977) [inline rllib/core/learner/tf/tf_learner.py:112]: Moved to base class. Possible b/c of new enhanced APIs.

File: rllib/core/learner/learner.py [hunk 4]
```diff
@@ -495,20 +613,97 @@ def register_metric(self, module_id: str, key: str, value: Any) -> None:
...
                 given `module_id`) to the actual values of these metrics. Values might
                 also be tensor vars (e.g. from within a traced tf2 function).
+                These will be automatically converted to numpy values.
         """
         for key, value in metrics_dict.items():
             self.register_metric(module_id, key, value)
 
-    def get_weights(self, module_ids: Optional[Set[str]] = None) -> Mapping[str, Any]:
-        """Returns the weights of the underlying MultiAgentRLModule.
+    def get_optimizer(
```
Comment (sven1977) [inline rllib/core/learner/learner.py:629]: New APIs to more easily retrieve registered optimizer by their module_ids and names AND to compile parameter sub-dictionaries that only contain param_refs -> params key/value-paris of a particular optimizer.

File: rllib/core/learner/learner.py [hunk 5]
```diff
@@ -306,147 +347,123 @@ def hps(self) -> LearnerHyperparameters:
...
+                device=self._device,
+            )
+            self._optimizer_lr_schedules[optimizer] = scheduler
+            # Set the optimizer to the current (first) learning rate.
+            self._set_optimizer_lr(
+                optimizer=optimizer,
+                lr=scheduler.get_current_value(),
+            )
 
-    def _configure_optimizers_for_module_helper(
```
Comment (sven1977) [inline rllib/core/learner/learner.py:347]: This method has become superflous due to the new register_optimizer API.

File: rllib/core/learner/learner.py [hunk 6]
```diff
@@ -306,147 +347,123 @@ def hps(self) -> LearnerHyperparameters:
...
         return self._hps
 
-    def configure_optimizers(self) -> ParamOptimizerPairs:
-        """Configures the optimizers for the Learner.
-
-        Do not override this method for your custom algorithms (which require certain
-        optimizers), but rather override the `self.configure_optimizers_for_module(
-        module_id=..)` method and return those optimizers from there that you need for
-        the given module.
+    def register_optimizer(
```
Comment (sven1977) [inline rllib/core/learner/learner.py:350]: New API allowing for more flexibility in users overriding `configure_optimizers` OR `configure_optimizers_for_module`.

**Comment by ollie-iterators** [comment]
Lint error:
rllib/core/testing/utils.py:7:1: F401 'ray.rllib.core.learner.learner.LearnerHyperparameters' imported but unused

File: rllib/core/learner/learner.py (hunk 6 above)
Comment (kouroshHakha) [reply rllib/core/learner/learner.py:350]: Very nice API change. I love it.

File: rllib/core/learner/learner.py [hunk 7]
```diff
@@ -495,20 +613,98 @@ def register_metric(self, module_id: str, key: str, value: Any) -> None:
...
+        named_optimizers = []
+        for full_registration_name in self._module_optimizers[module_id]:
+            optimizer = self._named_optimizers[full_registration_name]
+            # TODO (sven): How can we avoid registering optimziers under this
+            #  constructed `[module_id]_[optim_name]` format?
+            optim_name = full_registration_name[len(module_id) + 1 :]
+            named_optimizers.append((optim_name, optimizer))
+        return named_optimizers
+
+    def compile_param_dict_for_optimizer(
```
Comment (kouroshHakha) [inline rllib/core/learner/learner.py:680]: **super nit**: the name does not click with me right off the bat. Maybe `filter_param_dict_for_optimizer`?

File: rllib/core/learner/learner.py [hunk 8]
```diff
@@ -861,33 +1045,32 @@ def additional_update_for_module(
...
+
+        # Only cover the optimizer mapped to this particular module.
+        for optimizer_name, optimizer in self.get_optimizers_for_module(module_id):
+            # Only update this optimizer's lr, if a scheduler has been registered
+            # along with it.
+            if optimizer in self._optimizer_lr_schedules:
+                new_lr = self._optimizer_lr_schedules[optimizer].update(
+                    timestep=timestep
+                )
+                self._set_optimizer_lr(optimizer, lr=new_lr)
```
Comment (kouroshHakha) [inline rllib/core/learner/learner.py:1065]: **super nit**: Thinking about possible extensions coming from prospect users, I think we probably wanna also make the components of this method public API. like introducing `get_lr_scheduler(optimizer_name, optimizer)` and `set_optimizer_lr(optimizer, lr).

```
for optimizer_name, optimizer in self.get_optimizers_for_module(module_id):
            # Only update this optimizer's lr, if a scheduler has been registered
            # along with it.
            scheduler = get_lr_scheduler(optimizer_name, optimizer)
            if scheduler is not None:
                new_lr = scheduler.update(timestep=timestep)
                self.set_optimizer_lr(optimizer, lr=new_lr)
                # Make sure our returned results differentiate by optimizer name
                # (if not the default name).
                stats_name = LEARNER_RESULTS_CURR_LR_KEY
                if optimizer_name != DEFAULT_OPTIMIZER:
                    stats_name += "_" + optimizer_name
                results.update({stats_name: new_lr})
```

File: rllib/core/learner/learner.py [hunk 9]
```diff
@@ -1189,6 +1368,23 @@ def _make_module(self) -> MultiAgentRLModule:
         module = module.as_multi_agent()
         return module
 
+    def _check_registered_optimizer(
```
Comment (kouroshHakha) [inline rllib/core/learner/learner.py:1371]: I couldn't find where we call this function.

**Review by kouroshHakha — APPROVED** [review]
Everything looks good. I just notice you introduce an API here that is not used (i.e. `_check_registered_optimizer`)

**Comment by sven1977** [comment]
Thanks for the thorough review @kouroshHakha ! On this PR and the closed predecessor. I know it was a hell of a lot of work to go through these changes and your suggestions helped a ton to make the Learner API. Hopefully, this will get us into quieter waters now on Learner (unblocking Dreamer, etc..). 👍 💯 😃 
I'll fix the items you suggested and then merge.

File: rllib/core/learner/learner.py (hunk 7 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:680]: Yes, "filter" is much better. :)

Comment (sven1977) [reply rllib/core/learner/learner.py:680]: done

File: rllib/core/learner/learner.py (hunk 9 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:1371]: Totally, I erased the main call in `Learner.register_optimizer()` :) Fixed.

File: rllib/core/learner/learner.py [hunk 10]
```diff
@@ -483,7 +604,7 @@ def apply_gradients(self, gradients: ParamDict) -> None:
         """
 
     def register_metric(self, module_id: str, key: str, value: Any) -> None:
-        """Registers a single key/value metric pair for loss and gradient stats.
+        """Registers a single key/value metric pair for loss- and gradient stats.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/learner/learner.py:607]: Yes, as in: "loss stats and gradient stats" -> compress to -> "loss- and gradient stats"
